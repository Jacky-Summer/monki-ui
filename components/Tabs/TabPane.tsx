import React, { FC } from 'react'

export interface Props {
  tab: React.ReactNode
  disabled?: boolean
}

const TabPane: FC<Props> = ({ children }) => (
  <div className="mk-tab-panel">{children}</div>
)

export default TabPane
