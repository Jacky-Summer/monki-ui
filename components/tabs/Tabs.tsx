import React, { FC, FunctionComponentElement, MouseEvent, useState } from 'react'
import classNames from 'classnames'
import TabPane, { Props as TabPaneProps } from './TabPane'

export type TabsType = 'line' | 'card'

export interface TabsProps {
  className?: string
  defaultActiveKey?: number
  type?: TabsType
  onTabClick?: (selectedKey: number) => void
}

const RootTab: FC<TabsProps> = ({
  className,
  defaultActiveKey,
  type,
  onTabClick,
  children,
}) => {
  const [activeKey, setActiveKey] = useState(defaultActiveKey)
  const navClass = classNames('mk-tabs-nav', {
    'nav-line': type === 'line',
    'nav-card': type === 'card',
  })

  const handleClick = (e: MouseEvent, index: number, disabled: boolean | undefined) => {
    if (!disabled) {
      setActiveKey(index)
      if (onTabClick) {
        onTabClick(index)
      }
    }
  }

  const renderNavLinks = () =>
    React.Children.map(children, (child, index) => {
      const childElement = child as FunctionComponentElement<TabPaneProps>
      const { tab, disabled } = childElement.props
      const classes = classNames('mk-tabs-nav-item', {
        'is-active': activeKey === index,
        disabled,
      })

      return (
        <li
          className={classes}
          /* eslint-disable-next-line react/no-array-index-key */
          key={`nav-item-${index}`}
          onClick={e => {
            handleClick(e, index, disabled)
          }}
        >
          {tab}
        </li>
      )
    })

  const renderContent = () =>
    React.Children.map(children, (child, index) => {
      if (index === activeKey) {
        return child
      }
    })

  return (
    <div className={`mk-tabs ${className}`}>
      <ul className={navClass}>{renderNavLinks()}</ul>
      <div className="mk-tabs-content">{renderContent()}</div>
    </div>
  )
}

RootTab.defaultProps = {
  defaultActiveKey: 0,
  type: 'line',
}

export type ITabComponent = FC<TabsProps> & {
  TabPane: FC<TabPaneProps>
}

const Tabs = RootTab as ITabComponent
Tabs.TabPane = TabPane

export default Tabs
