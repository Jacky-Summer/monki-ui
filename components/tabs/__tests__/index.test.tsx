import React from 'react'
import { mount } from 'enzyme'
import Tabs, { TabsProps } from '../Tabs'

const { TabPane } = Tabs

const testProps: TabsProps = {
  defaultActiveKey: 1,
  onTabClick: jest.fn(),
}

const generateTabs = (props: TabsProps = {}) => (
  <Tabs {...testProps}>
    <TabPane tab="Tab 1" disabled>
      Content of Tab Pane 1
    </TabPane>
    <TabPane tab="Tab 2">Content of Tab Pane 2</TabPane>
    <TabPane tab="Tab 3">Content of Tab Pane 3</TabPane>
  </Tabs>
)

describe('Tabs component', () => {
  it('should render the correct default Tabs', () => {
    const wrapper = mount(generateTabs())
    const activeElement = wrapper.find('.mk-tabs-nav-item').at(1)
    expect(wrapper.find('.mk-tabs-nav').hasClass('nav-line'))
    expect(activeElement.hasClass('is-active')).toBeTruthy()
    expect(activeElement.text()).toEqual('Tab 2')
    expect(wrapper.find('.mk-tab-panel').text()).toEqual('Content of Tab Pane 2')
    expect(wrapper).toMatchSnapshot()
  })

  it('click tabItem should switch to content', () => {
    const wrapper = mount(generateTabs())
    wrapper
      .find('.mk-tabs-nav-item')
      .at(2)
      .simulate('click')

    expect(testProps.onTabClick).toHaveBeenCalled()
    expect(
      wrapper
        .find('.mk-tabs-nav-item')
        .at(2)
        .hasClass('is-active'),
    ).toBeTruthy()
    expect(
      wrapper
        .find('.mk-tabs-nav-item')
        .at(1)
        .hasClass('is-active'),
    ).toBeFalsy()
  })

  it('click disabled tabItem should not works', () => {
    const wrapper = mount(generateTabs())
    wrapper
      .find('.mk-tabs-nav-item')
      .at(0)
      .simulate('click')

    expect(
      wrapper
        .find('.mk-tabs-nav-item')
        .at(0)
        .hasClass('is-active'),
    ).toBeFalsy()
    expect(
      wrapper
        .find('.mk-tabs-nav-item')
        .at(1)
        .hasClass('is-active'),
    ).toBeTruthy()
  })
})
