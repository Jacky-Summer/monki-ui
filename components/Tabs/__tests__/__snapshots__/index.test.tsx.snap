// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs component should render the correct default Tabs 1`] = `
<RootTab
  defaultActiveKey={1}
  onTabClick={[MockFunction]}
  type="line"
>
  <div
    className="mk-tabs undefined"
  >
    <ul
      className="mk-tabs-nav nav-line"
    >
      <li
        className="mk-tabs-nav-item disabled"
        key="nav-item-0/.0"
        onClick={[Function]}
      >
        Tab 1
      </li>
      <li
        className="mk-tabs-nav-item is-active"
        key="nav-item-1/.1"
        onClick={[Function]}
      >
        Tab 2
      </li>
      <li
        className="mk-tabs-nav-item"
        key="nav-item-2/.2"
        onClick={[Function]}
      >
        Tab 3
      </li>
    </ul>
    <div
      className="mk-tabs-content"
    >
      <TabPane
        key=".1"
        tab="Tab 2"
      >
        <div
          className="mk-tab-panel"
        >
          Content of Tab Pane 2
        </div>
      </TabPane>
    </div>
  </div>
</RootTab>
`;
